File: ЛР5 Полиморфизм С++/MyClasses.cpp
#include "MyClasses.h"

int Characters::strlength(string str) {
    int len = 0;
    while (str[len] != '\0') {
        len++;
    }
    return len;
}
int Numbers::strlength(string str) {
    int len = 0;
    while (str[len] != '\0') {
        len++;
    }
    return len;
}


Characters::Characters(string s1)
{
    s = s1;
}

int Characters::str_len()
{
    return strlength(s);
}
string Characters::char_replace()
{
    int counter = 0;
    for (int i = 0; i < strlength(s); i++) { if (s[i] == '#') counter++; }
    string arr = s;
    for (int i = 0; i < counter; i++) arr += '1';
    for (int i = 0, j = 0; i < strlength(s); i++, j++)
    {
        if (s[i] != '#')  arr[j] = s[i];
        else
        {
            arr[j] = '!';
            arr[j + 1] = '!';
            j++;
        }
    }
    s = arr;
    return s;
}

Numbers::Numbers(string s1)
{
    int counter = 0;
    for (int i = 0; i < strlength(s1); i++) { if (isdigit(s1[i])) counter++; }
    string arr = "";
    for (int i = 0; i < counter; i++) arr += '1';
    for (int i = 0, j = 0; j < strlength(s1); i++, j++)
    {
        if (isdigit(s1[j])) arr[i] = s1[j];
        else i--;
    }
    s = arr;
}
int Numbers::str_len()
{
    return strlength(s);
}
string Numbers::char_replace()
{
    int counter = 0;
    for (int i = 0; i < strlength(s); i++) { if (s[i] == '3') counter++; }
    string arr = s;
    for (int i = 0; i < counter; i++) arr += '1';
    for (int i = 0, j = 0; i < strlength(s); i++, j++)
    {
        if (s[i] != '3')  arr[j] = s[i];
        else
        {
            arr[j] = '1';
            arr[j + 1] = '1';
            j++;
        }
    }
    s = arr;
    return s;
}

File: ЛР5 Полиморфизм С-sharp/MyClasses.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace ConsoleApp2
{
    abstract class String
    {
        virtual public int str_len() { return 0; }
        virtual public string char_replace() { return ""; }
    }

    class Characters : String
    {
        private string s;

        public Characters(string s1)
        {
            s = s1;
        }

        override public int str_len()
        {
            return s.Length;
        }
        override public string char_replace()
        {
            int counter = 0;
            for (int i = 0; i < s.Length; i++) { if (s[i] == '#') counter++; }
            char[] arr = new char[s.Length];
            Array.Resize(ref arr, s.Length + counter);
            for (int i = 0, j = 0; i < s.Length; i++, j++)
            {
                if (s[i] != '#') arr[j] = s[i];
                else
                {
                    arr[j] = '!';
                    arr[j + 1] = '!';
                    j++;
                }
            }
            s = new string(arr);
            return s;
        }
    }

    class Numbers : String
    {
        private string s;

        public Numbers(string s1)
        {
            int counter = 0;
            for (int i = 0; i < s1.Length; i++) { if (Char.IsDigit(s1[i])) counter++; }
            char[] arr = new char[counter];
            for (int i = 0, j = 0; j < s1.Length; i++, j++)
            {
                if (Char.IsDigit(s1[j])) arr[i] = s1[j];
                else i--;
            }
            s = new string(arr);
        }

        override public int str_len()
        {
            return s.Length;
        }
        override public string char_replace()
        {
            int counter = 0;
            for (int i = 0; i < s.Length; i++) { if (s[i] == '3') counter++; }
            char[] arr = new char[s.Length];
            Array.Resize(ref arr, s.Length + counter);
            for (int i = 0, j = 0; i < s.Length; i++, j++)
            {
                if (s[i] != '3') arr[j] = s[i];
                else
                {
                    arr[j] = '1';
                    arr[j + 1] = '1';
                    j++;
                }
            }
            s = new string(arr);
            return s;
        }
    }
}
